docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,364 @@
+
+##### Task Config
+
+We do not need any config for this task. When creating an instance of `EtcdOperatorTask` for this scenario, `.spec.config` will be set to nil (unset).
+
+##### Pre-Conditions
+
+* There should not be a `on-demand snapshot-compaction` task already running for the same etcd cluster.
+
+>Note: `on-demand snapshot-compaction` runs as a separate job in a seprate pod, hence it doesn't depends on health of etcd cluster members or any other conditions.
+
+#### Trigger on-demand full/delta snapshot
+
+`Etcd` custom resource provides an ability to set [FullSnapshotSchedule](https://github.com/gardener/etcd-druid/blob/master/api/v1alpha1/types_etcd.go#L158) which is currently defauled to run once in 24 hrs. [DeltaSnapshotPeriod](https://github.com/gardener/etcd-druid/blob/master/api/v1alpha1/types_etcd.go#L171) is also made configurable which defines the duration after which a delta snapshot will be taken.
+If operator does not wish to wait for the scheduled full/delta snapshot, he/she can trigger an on-demand(out-of-schedule) full/delta snapshot on etcd cluster which will be taken by `leading-backup-restore`.
+
+##### Possible scenarios
+
+* [Shoot hibernation](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_hibernate.md): Every etcd cluster incurs an inherent cost of preserving the volumes even when the shoot is in hibernation state. However it is possible to save costs by invoking this task to take a full snapshot before deleting the volumes.
+* [Control Plane Migration](https://github.com/gardener/gardener/blob/master/docs/proposals/07-shoot-control-plane-migration.md): In [gardener](https://github.com/gardener/gardener) a seed cluster control plane can be moved to another seed cluster. To prevent data loss and faster restoration of the etcd cluster in the target seed, a full snapshot can be triggered for the etcd cluster in the source seed.
+* A on-demand full snapshot can be triggered if scheduled snapshot fails due to any reason.
+
+##### Task Config
+
+```go
+// SnapshotType can be full or delta snapshot
+SnapshotType string `json:"snapshotType"`
+const (
+  FullSnapshot SnapshotType = "full-snapshot"
+  DeltaSnapshot SnapshotType = "delta-snapshot"
+)
+
+type OnDemandSnapshotTaskConfig struct {
+  // Type of on-demand snapshot.
+  SnapshotType snapshotType `json:"snapshotType"`
+}
+```
+
+##### Pre-Conditions
+
+* Etcd cluster should have a quorum.
+* There should not already be a `on-demand snapshot` task running with the same `SnapshotType` for the same etcd cluster.
+
+#### Trigger on-demand maintenance of etcd cluster
+
+Operator can trigger on-demand maintenance of etcd cluster which includes operations like [etcd compaction](https://etcd.io/docs/v3.5/op-guide/maintenance/#history-compaction-v3-api-key-value-database), [etcd defragmentation](https://etcd.io/docs/v3.2/op-guide/maintenance/#defragmentation) etc.
+
```

Comment:
Does this also perform the `on-demand full` after the `compact` & `defrag`, if not then it would be better to include that as then we don't have to run `on-demand compaction` for some time after this.

Reply:
> Does this also perform the on-demand full after the compact & defrag,

As the task name suggest `on-demand maintenance of etcd cluster`, it just orchestrate the maintenance of cluster out of maintenance window, it doesn't trigger the `on-demand full snapshot`. `out-of-band` task should do what its required to do, nothing more than that.

>if not then it would be better to include that as then we don't have to run on-demand compaction for some time after this

Main reason of introducing `on-demand snapshot compaction` is it will run as a separate job without affecting or depending on health of etcd cluster.
Now, if some operator is invoking the `on-demand maintenance` task, it means that etcd cluster is already heavily loaded and now, we are saying after `maintenance`, also take full-snapshot which will also cause more load on running etcd cluster. I'm not in favour of that.
However, still somebody wants to take full-snapshot after `on-demand maintenance` then he/she can invoke `on-demand snapshot` task to take it

Reply:
I agree with your POV of load on the cluster. Whatever I said only made sense for me because of this [comment](https://github.com/gardener/etcd-druid/pull/757#discussion_r1552933574). These both were supposed to go hand-in-hand. But now that you detailed on why the tasks need to be separate so as to study the reason for certain events, I'm also in favour of having tasks independent, as there is no need to complicate things. Thanks for explaining.